### PR TITLE
feat(chain): add `get` and `range` methods to `CheckPoint`

### LINKED ISSUE #1354: The `TxGraph::missing_heights` implementation is problematic.
The logic of `missing_heights` is explained clearly in this comment:

```rust
// Usually, if a height of a tx anchor is missing from the chain, we would want to return
// this height in the iterator. The exception is when the tx is confirmed in chain. All the
// other missing-height anchors of this tx can be skipped.
```

However, if a transaction is confirmed in a block that is about to turn stale (get reorged out), we
still want to fetch the new block hash. I suspect this is where our "transaction is never confirmed"
bug comes from.

**My proposed solution:**

Change the exception to:

```rust
// ... The exception is when the tx is confirmed in chain with a depth greater than the provided
// `max_reorg_depth`.
```

And we change the signature of the method to be:

```rust
pub fn missing_heights<'a>(
    &'a self,
    chain: &'a LocalChain,
    max_reorg_depth: u32,
) -> impl Iterator<Item = u32> + 'a { todo!() }
```

With some nice documentation for `max_reorg_depth`.

**An alternative solution:**

Remove the stated "exception". We can continue to consider a height "missing" when there is a
floating anchor at that height (anchor that does not point to a block in the best chain).

The downside of this approach is that there are situations where certain heights will always be
considered "missing". I think this is somewhat acceptable as reorgs are rare and this situation only
happens due to reorgs.

Because of the behavior change, if we do this solution, we should rename the method to
`floating_anchor_heights`.

**How to reproduce with `bdk_esplora`:**

The following should be written as a test.

* Wallet receives money via transaction (A).
* Transaction gets confirmed at confirmation height (h).
* (Sync wallet with `bdk_esplora`),
* A reorg happens, and (A) gets re-confirmed at height (h') where `h' >= h`.
* (Sync wallet with `bdk_esplora`).
* At this point onwards, transaction (A) is forever stuck as unconfirmed!

**Going forward:**

The test should be written first while we wait for the discussion on which solution to do matures.

### PR DESCRIPTION
**Pull request by evanlinjin**
Partially fixes #1354 

### Description

These methods allow us to query for checkpoints contained within the linked list by height and height range. This is useful to determine checkpoints to fetch for chain sources without having to refer back to the `LocalChain`.

Currently this is not implemented efficiently, but in the future, we will change `CheckPoint` to use a skip list structure.

### Notes to the reviewers

Please refer to the conversation in #1354 for more details. In summary, both `TxGraph::missing_heights` and `tx_graph::ChangeSet::missing_heights_from` are problematic in their own ways. Additionally, it's a better API for chain sources if we only need to lock our receiving structures twice (once to obtain initial state and once for applying the update). Instead of relying on methods such as `EsploraExt::update_local_chain` to get relevant checkpoints, we can use these query methods instead. This allows up to get rid of `::missing_heights`-esc methods and remove the need to lock receiving structures in the middle of the sync.

### Changelog notice

* Added `get` and `range` methods to `CheckPoint` (and in turn, `LocalChain`). This simulates an API where we have implemented a skip list of checkpoints (to implement in the future). This is a better API because we can query for any height or height range with just a checkpoint tip instead of relying on a separate checkpoint index (which needs to live in `LocalChain`).
* Changed `LocalChain` to have a faster `Eq` implementation. We now maintain an xor value of all checkpoint block hashes. We compare this xor value to determine whether two chains are equal.
* Added `PartialEq` implementation for `CheckPoint` and `local_chain::Update`.

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [x] I've added tests for the new feature
* [x] I've added docs for the new feature

### CODE REVIEW

File: crates/chain/src/local_chain.rs [hunk 1]
```diff
@@ -148,6 +149,34 @@ impl CheckPoint {
...
+
+    /// Find checkpoint at `height`.
+    ///
+    /// Returns `None` if checkpoint at `height` does not exist`. If you are querying for multiple
+    /// heights, it will be more efficient to use [`query_multi`].
+    ///
+    /// [`query_multi`]: CheckPoint::query_multi
+    pub fn query(&self, height: u32) -> Option<Self> {
+        self.iter()
+            .take_while(|cp| cp.height() >= height)
```
Comment (LLFourn) [inline crates/chain/src/local_chain.rs]: No need for this `take_while`.

File: crates/chain/src/local_chain.rs [hunk 2]
```diff
@@ -148,6 +149,34 @@ impl CheckPoint {
...
+            .find(|cp| cp.height() == height)
+    }
+
+    /// Find checkpoints of the given `heights`.
+    ///
+    /// This returns an iterator which iterates over the requested heights and whether a
+    /// corresponding checkpoint is found.
+    pub fn query_multi(&self, heights: BTreeSet<u32>) -> impl Iterator<Item = (u32, Option<Self>)> {
```
Comment (LLFourn) [inline crates/chain/src/local_chain.rs]: Seems like a premature optimization. I understand that this is faster than calling query several times but the way to speed `query` up is with a skip list.

**Review by LLFourn — COMMENTED** [review]
I'm slightly concerned we're introducing these methods without using them anywhere. Especially `query_multi` which I don't think needs to exist yet. I suggest cutting `query_multi`.

File: crates/chain/src/local_chain.rs (hunk 1 above)
Comment (evanlinjin) [reply crates/chain/src/local_chain.rs]: I think we need this. If we don't have this and we don't have a direct hit, we will iterate the entire chain.

Comment (LLFourn) [reply crates/chain/src/local_chain.rs]: Ah good point thanks! (comment that)

File: crates/chain/src/local_chain.rs (hunk 2 above)
Comment (evanlinjin) [reply crates/chain/src/local_chain.rs]: Even with a skip list, this will be faster? I also see this as a convenience method.

Comment (LLFourn) [reply crates/chain/src/local_chain.rs]: > Even with a skip list, this will be faster? This is also a convenience method, like super convenient. I honestly think this is worth it.

Things that are faster are not necessarily better (if done before we have evidence that we need faster). Can we add the convenience method at the point that we need the convenience?

I think in general this PR doesn't really belong at this point because we don't get to see how we actually use the API. It seems like a simple thing to do as part of a larger PR that deletes the BTreeMap from LocalChain and uses checkpoint queries instead. Can we just make that this PR?

File: crates/chain/src/local_chain.rs (hunk 1 above)
Comment (evanlinjin) [reply crates/chain/src/local_chain.rs]: Comment added.

File: crates/chain/src/local_chain.rs (hunk 2 above)
Comment (evanlinjin) [reply crates/chain/src/local_chain.rs]: I realized that you are right. I think it is a great policy to only introduce convenience methods if they are used. `CheckPoint::query_multi` turns out to be not very useful. However, I introduced `CheckPoint::query_from` which will return `Some` if there is a checkpoint at `height` or greater. This is a sufficient replacement for `BTreeMap::<u32, BlockHash>::range` which is used in `bdk::Wallet::insert_tx`.

File: crates/chain/src/local_chain.rs [hunk 3]
```diff
@@ -205,18 +224,39 @@ pub struct Update {
 #[derive(Debug, Clone)]
 pub struct LocalChain {
     tip: CheckPoint,
-    index: BTreeMap<u32, BlockHash>,
+    xor: BlockHashXor,
```
Comment (LLFourn) [inline crates/chain/src/local_chain.rs]: I think this is a bit premature. I don't know that we need really fast equality for anything.  We didn't have fast equality before. If we were to do it, I'd just do it in `CheckPoint`.

Comment (evanlinjin) [reply crates/chain/src/local_chain.rs]: > If we were to do it, I'd just do it in `CheckPoint`.

Do you mean implement fast equality in `CheckPoint`? So every checkpoint will have an `xor` field?

Comment (evanlinjin) [reply crates/chain/src/local_chain.rs]: Yes, I agree that adding the xor right now is premature.

File: crates/chain/src/local_chain.rs [hunk 4]
```diff
@@ -34,6 +34,20 @@ struct CPInner {
     prev: Option<Arc<CPInner>>,
 }
 
+impl PartialEq for CheckPoint {
+    fn eq(&self, other: &Self) -> bool {
```
Comment (LLFourn) [inline crates/chain/src/local_chain.rs:39]: :wrench: https://doc.rust-lang.org/core/iter/trait.Iterator.html#method.eq

I think it can be implemented with this.

**Review by LLFourn — APPROVED** [review]
ACK aa84ecf9d0a6f37e56c6fbebe3a2370f22f8acfd

File: crates/chain/tests/test_local_chain.rs [hunk 5]
```diff
@@ -528,6 +528,45 @@ fn checkpoint_from_block_ids() {
...
+    ];
+
+    for t in test_cases.into_iter() {
+        let tip = t.chain.tip();
+        for h in t.query_range.0..=t.query_range.1 {
+            let query_result = tip.query(h);
+            let exp_hash = t.chain.query(h).map(|cp| cp.hash());
+            match query_result {
+                Some(cp) => {
+                    assert_eq!(Some(cp.hash()), exp_hash);
```
Comment (ValuedMammal) [inline crates/chain/tests/test_local_chain.rs:568]: It kind of looks like we're getting the checkpoint hash twice here from the same query result.

File: crates/chain/src/local_chain.rs (hunk 3 above)
Comment (LLFourn) [reply crates/chain/src/local_chain.rs]: > > If we were to do it, I'd just do it in `CheckPoint`.
> 
> Do you mean implement fast equality in `CheckPoint`? So every checkpoint will have an `xor` field?

Yes. As discussed in #1380 we probably do want to do this for `CheckPoint` in the future.

File: crates/chain/src/local_chain.rs (hunk 4 above)
Comment (ValuedMammal) [reply crates/chain/src/local_chain.rs:39]: Would that not cause a recursion though?

Comment (LLFourn) [reply crates/chain/src/local_chain.rs:39]: Oh interesting point. We still want to map to the `BlockId` so the eq used in the iterator is just `BlockId`'s. That should be fine!

File: crates/chain/tests/test_local_chain.rs [hunk 6]
```diff
@@ -528,6 +528,45 @@ fn checkpoint_from_block_ids() {
...
+                    assert_eq!(cp.height(), h);
+                }
+                None => assert!(query_result.is_none()),
```
Comment (notmandatory) [inline crates/chain/tests/test_local_chain.rs]: 🔧 Since this match arm already told us `query_result` is `None` did you mean to do this instead?
```suggestion
                None => assert!(exp_hash.is_none()),
```

File: crates/chain/tests/test_local_chain.rs (hunk 5 above)
Comment (notmandatory) [reply crates/chain/tests/test_local_chain.rs:568]: I think this is OK, I read this line as saying: if the `query_result` is some checkpoint, then the hash of that checkpoint must equal the expected hash. 

⛏️ On the other hand I don't see how this test could ever fail since both `tip.query(h)` and `t.chain.query(h)` call `Checkpoint::query()` with `h`.

**Review by notmandatory — APPROVED** [review]
ACK aa84ecf9d0a6f37e56c6fbebe3a2370f22f8acfd

Appreciate the focused scope for this series of chain PRs, much easier to review. I have one small test code suggestion but otherwise looks good to me.

File: crates/chain/tests/test_local_chain.rs (hunk 5 above)
Comment (evanlinjin) [reply crates/chain/tests/test_local_chain.rs:568]: I realized how I wrote this was wrong. We shouldn't be using `query` for both what is to be checked and what we are checking against.

File: crates/chain/tests/test_local_chain.rs (hunk 6 above)
Comment (evanlinjin) [reply crates/chain/tests/test_local_chain.rs]: Fixed.

File: crates/chain/tests/test_local_chain.rs (hunk 5 above)
Comment (evanlinjin) [reply crates/chain/tests/test_local_chain.rs:568]: > I realized how I wrote this was wrong. We shouldn't be using `query` for both what is to be checked and what we are checking against.

Fixed.

File: crates/chain/src/local_chain.rs (hunk 4 above)
Comment (evanlinjin) [reply crates/chain/src/local_chain.rs:39]: Good point @ValuedMammal. I'll map the iterator to block id first.

Comment (evanlinjin) [reply crates/chain/src/local_chain.rs:39]: Updated to use `Iterator::eq`.

**Review by notmandatory — APPROVED** [review]
re-ACK 2b006da9042156b6ad9ce5a4237fc24973d0d72a

**Comment by evanlinjin** [comment]
I would hold off merging this until this discussion reaches a consensus: https://github.com/bitcoindevkit/bdk/pull/1380#discussion_r1540554140

**Comment by evanlinjin** [comment]
I have agreed with @LLFourn's suggestion of implementing `CheckPoint::range` and have updated the PR to implement it. Refer to the conversation here: https://github.com/bitcoindevkit/bdk/pull/1380#discussion_r1547182439

This PR is ready for review/merge.

**Comment by evanlinjin** [comment]
Ahhh I probably still need to add a few tests.

File: crates/chain/src/local_chain.rs [hunk 7]
```diff
@@ -148,6 +158,39 @@ impl CheckPoint {
...
     }
+
+    /// Query for checkpoint at `height`.
+    ///
+    /// Returns `None` if checkpoint at `height` does not exist`.
+    pub fn query(&self, height: u32) -> Option<Self> {
+        self.iter()
+            // optimization to avoid iterating the entire chain if we do not get a direct hit
+            .take_while(|cp| cp.height() >= height)
+            .find(|cp| cp.height() == height)
```
Comment (jirijakes) [inline crates/chain/src/local_chain.rs]: Perhaps opportunity to reuse `range`, since `query` is its special case? It seems that even performance-wise it would not be worse.

```suggestion
	self.range(height..=height).next()
```

File: crates/chain/src/local_chain.rs [hunk 8]
```diff
@@ -148,6 +158,39 @@ impl CheckPoint {
     pub fn iter(&self) -> CheckPointIter {
         self.clone().into_iter()
     }
+
+    /// Query for checkpoint at `height`.
+    ///
+    /// Returns `None` if checkpoint at `height` does not exist`.
+    pub fn query(&self, height: u32) -> Option<Self> {
```
Comment (jirijakes) [inline crates/chain/src/local_chain.rs]: The name feels hard to discover. If I am interested in finding a checkpoint at a _height_, I would first look for a method containing _height_, something like `find_at_height()`, `find_by_height()`, `get_at_height()`, `at_height()` or so.

Similarly `range` and the equivalent methods on `LocalChain`.

Comment (evanlinjin) [reply crates/chain/src/local_chain.rs]: What about `query_height`? cc @LLFourn

Comment (LLFourn) [reply crates/chain/src/local_chain.rs]: I'd go with just `.get_hash` or better yet just `get`. `query` is not a good name indeed.

File: crates/chain/src/local_chain.rs [hunk 9]
```diff
@@ -148,6 +158,39 @@ impl CheckPoint {
...
+        self.iter()
+            // optimization to avoid iterating the entire chain if we do not get a direct hit
+            .take_while(|cp| cp.height() >= height)
+            .find(|cp| cp.height() == height)
+    }
+
+    /// Iterate checkpoints over a height range.
+    ///
+    /// Note that we always iterate checkpoints in reverse height order (iteration starts at tip
```
Comment (LLFourn) [inline crates/chain/src/local_chain.rs:170]: :pick: there's no need to start this sentence with `Note that`. It doesn't start at the tip it starts at the highest height within the range.

The implementation looks correct fwiw. I still feel like we are adding methods like this before we actually need them. My bias is to keep it for now since you've done it but we really have to stop adding these kinds of methods.

Comment (evanlinjin) [reply crates/chain/src/local_chain.rs:170]: This is needed though, like we're actually using it

File: crates/chain/src/local_chain.rs (hunk 8 above)
Comment (evanlinjin) [reply crates/chain/src/local_chain.rs]: Let's go with `.get`.

File: crates/chain/src/local_chain.rs [hunk 10]
```diff
@@ -202,21 +245,19 @@ pub struct Update {
...
-    index: BTreeMap<u32, BlockHash>,
-}
-
-impl PartialEq for LocalChain {
-    fn eq(&self, other: &Self) -> bool {
-        self.index == other.index
-    }
 }
 
+// TODO: Figure out whether we can get rid of this
```
Comment (ValuedMammal) [inline crates/chain/src/local_chain.rs]: You could make this be the implementation for the `LocalChain::blocks` function and have it return `BTreeMap<u32, BlockHash>`, possibly even taking an extra `range: impl RangeBounds<u32>` parameter.

File: crates/chain/src/local_chain.rs [hunk 11]
```diff
@@ -1,6 +1,8 @@
 //! The [`LocalChain`] is a local implementation of [`ChainOracle`].
 
 use core::convert::Infallible;
+use core::ops::RangeBounds;
+use core::u32;
```
Comment (ValuedMammal) [inline crates/chain/src/local_chain.rs]: Is this `core::u32` needed?

**Review by ValuedMammal — COMMENTED** [review]
Overall looks good.

`get` sounds like a better name than `query`. My other suggestion would have been to call it `find`, since that is how it's implemented.

Interesting that `LocalChain` now delegates virtually everything to the checkpoint. Curious to see if the structure of `LocalChain` becomes more relevant when skip list is involved.

File: crates/chain/src/local_chain.rs (hunk 9 above)
Comment (LLFourn) [reply crates/chain/src/local_chain.rs:170]: I see one place it's being used where [as we've discussed](https://github.com/bitcoindevkit/bdk/pull/1380#discussion_r1540554140) it's not necessary.

File: crates/bdk/src/wallet/mod.rs [hunk 12]
```diff
@@ -1128,18 +1128,14 @@ impl<D> Wallet<D> {
                 // anchor tx to checkpoint with lowest height that is >= position's height
                 let anchor = self
                     .chain
-                    .blocks()
                     .range(height..)
-                    .next()
+                    .last()
```
Comment (evanlinjin) [inline crates/bdk/src/wallet/mod.rs:1131]: @LLFourn `CheckPoint::range` is also used here.

Comment (LLFourn) [reply crates/bdk/src/wallet/mod.rs:1131]: Thanks. I missed that one and had forgot about the API of this function. On a side note this is interesting because we are anchoring a tx not to the block that it's in. So even if we made a rule that esplora/electrum etc must tell us about the block that the tx is in we'd still need the concept of anchoring if we want to keep this API.

**Comment by LLFourn** [comment]
ACK cdbc131ef68182a4317beb145232ec8e87d6299c

**Review by LLFourn — APPROVED** [review]

File: crates/chain/src/local_chain.rs (hunk 11 above)
Comment (evanlinjin) [reply crates/chain/src/local_chain.rs]: I blame my autocomplete engine.

Comment (evanlinjin) [reply crates/chain/src/local_chain.rs]: Removed

File: crates/chain/src/local_chain.rs (hunk 7 above)
Comment (evanlinjin) [reply crates/chain/src/local_chain.rs]: Changed.

File: crates/chain/src/local_chain.rs (hunk 10 above)
Comment (evanlinjin) [reply crates/chain/src/local_chain.rs]: This is an interesting idea! However, I like to use @LLFourn's policy when it comes to introducing helper methods/structures. We should only introduce such things if we introduce how/where it is used within the PR. This ensures the usefulness of the things we are introducing.

I've decided to remove this `impl` completely as it was not being used.

File: crates/chain/src/local_chain.rs (hunk 11 above)
Comment (evanlinjin) [reply crates/chain/src/local_chain.rs]: Removed.

File: crates/chain/src/local_chain.rs (hunk 9 above)
Comment (evanlinjin) [reply crates/chain/src/local_chain.rs:170]: Conversation continues here: https://github.com/bitcoindevkit/bdk/pull/1369#discussion_r1550753230

**Comment by evanlinjin** [comment]
I've addressed all review comments and removed the `From<LocalChain>` impl for `BTreeMap<u32, BlockHash>` as it wasn't being used.

File: crates/chain/src/local_chain.rs [hunk 13]
```diff
@@ -276,39 +299,30 @@ impl LocalChain {
...
+        Ok(Self { tip })
     }
 
     /// Constructs a [`LocalChain`] from a [`BTreeMap`] of height to [`BlockHash`].
     ///
     /// The [`BTreeMap`] enforces the height order. However, the caller must ensure the blocks are
     /// all of the same chain.
     pub fn from_blocks(blocks: BTreeMap<u32, BlockHash>) -> Result<Self, MissingGenesisError> {
-        if !blocks.contains_key(&0) {
+        if blocks.get(&0).is_none() {
```
Comment (jirijakes) [inline crates/chain/src/local_chain.rs:321]: Just curious: why this change? Original looked a bit better readable.

File: crates/chain/src/local_chain.rs [hunk 14]
```diff
@@ -542,33 +550,41 @@ impl LocalChain {
...
+            // The checkpoint below the earliest checkpoint to remove will be the new tip.
+            Some(Some(new_tip)) => new_tip,
+            // If there is no checkpoint below the earliest checkpoint to remove, it means the
+            // "earliest checkpoint to remove" is the genesis block. We disallow removing the
+            // genesis block.
+            Some(None) => return Err(MissingGenesisError),
+            // If there is nothing to remove, we return an empty changeset.
+            None => return Ok(ChangeSet::default()),
+        };
+        Ok(changeset)
```
Comment (jirijakes) [inline crates/chain/src/local_chain.rs:575]: Not too important, but it took me a while to understand what changeset is returned when and what is when assigned to `self.tip` and when it is left untouched.

Perhaps?

```suggestion
        match remove_from.map(|cp| cp.prev()) {
            // The checkpoint below the earliest checkpoint to remove will be the new tip.
            Some(Some(new_tip)) => {
                self.tip = new_tip;
                Ok(changeset)
            }
            // If there is no checkpoint below the earliest checkpoint to remove, it means the
            // "earliest checkpoint to remove" is the genesis block. We disallow removing the
            // genesis block.
            Some(None) => Err(MissingGenesisError),
            // If there is nothing to remove, we return an empty changeset.
            None => Ok(ChangeSet::default()),
        }
```

**Review by jirijakes — APPROVED** [review]
The PR helped me better understand some internals. Still learning.

Only have one question and one minor suggestion (not too important).

ACK 6d78f0e0fbaf7b25499ed671fdd2c1266a0d47c7

File: crates/chain/src/local_chain.rs (hunk 13 above)
Comment (evanlinjin) [reply crates/chain/src/local_chain.rs:321]: `.contains_key` no longer exists because `blocks` is a different type now.

**Review by LLFourn — APPROVED** [review]
ACK 53942cced492138174638a8087b08f643a8d41ad

File: crates/chain/src/local_chain.rs (hunk 14 above)
Comment (evanlinjin) [reply crates/chain/src/local_chain.rs:575]: Your suggestion is definitely better, but I'll ignore it now because I want this merged and rebasing is annoying 😅
